dwave/cloud/regions.py

```diff
@@ -0,0 +1,183 @@
+            connection parameters in full. Raw config (full or partial) loaded
+            from a file/env with :meth:`~dwave.cloud.config.loaders.load_config`
+            can be passed in as a :class:`dict` and it will be validated as
+            :class:`~dwave.cloud.config.models.ClientConfig`. A string ``config``
+            value is interpreted as ``metadata_api_endpoint``.
+        refresh:
+            Force regions cache refresh.
+        maxage:
+            Maximum allowed age of cached regions metadata.
+        no_cache:
+            Don't use cache.
+
+    Returns:
+        List of :class:`~dwave.cloud.api.models.Region` objects, each describing
+        one region.
+
+    .. versionadded:: 0.11.0
+        Added :func:`.get_regions`.
+
+    """
+    if isinstance(config, str):
+        config = ClientConfig(metadata_api_endpoint=config)
+    elif isinstance(config, dict):
+        config = validate_config_v1(config)
+    elif config is None:
+        config = ClientConfig()
+    elif not isinstance(config, ClientConfig):
+        raise TypeError(f"'config' type {type(config).__name__!r} not supported")
+
+    # make sure we don't cache by noise in config
+    cache_key = (config.metadata_api_endpoint, config.cert, config.headers,
+                 config.proxy, config.permissive_ssl,
+                 config.request_retry, config.request_timeout)
+
+    fetch = partial(_fetch_available_regions,
+        cache_key=cache_key, config=config, refresh_=refresh, maxage_=maxage)
+
+    if no_cache:
+        fetch = cached.disabled()(fetch)
+
+    try:
+        obj = fetch()
+    except api.exceptions.RequestError as exc:
+        logger.debug("Metadata API unavailable", exc_info=True)
+        raise ValueError(
+            f"Metadata API unavailable at {config.metadata_api_endpoint!r}")
+
+    regions = TypeAdapter(List[api.models.Region]).validate_python(obj)
+    return regions
+
+
+def _infer_leap_api_endpoint(solver_api_endpoint: str,
+                             region_code: Optional[str] = None) -> str:
+    # shim until metadata api includes leap endpoint in region data
+
+    parts = urlsplit(solver_api_endpoint)
+
+    # infer region code if necessary
+    if not region_code:
+        if m := re.match(r'([a-z]+-[a-z]+-\d+)\.', parts.netloc, re.I):
+            region_code = m.group(1)
+        else:
+            region_code = DEFAULT_REGION
+
+    # update path
+    path = urlsplit(DEFAULT_LEAP_API_ENDPOINT).path
+
+    # strip region from netloc
+    netloc = parts.netloc
+    if parts.netloc.startswith(f"{region_code}."):
+        netloc = parts.netloc[len(region_code)+1:]
+
+    return parts._replace(netloc=netloc, path=path).geturl()
+
+
+def resolve_endpoints(config: ClientConfig, *, inplace: bool = False) -> ClientConfig:
+    """Use region and endpoint from config to resolve all config endpoints.
+
+    Explicit endpoint will override the region (i.e. region extension is
+    backwards-compatible).
```

Comment:
Not sure the comment about backwards compatibility should be in docstring -- it will grow stale. Perhaps as a comment?

Reply:
I think it's good to raise awareness of this as a backwards-compatible feature, only for users be aware when we change it. :slightly_smiling_face:

Reply:
Maybe add to the release notes then?